### ISSUE TITLE
Paragraph with many child runs causes stack overflow

### ISSUE THREAD

**Issue by tolrtd** [mwilliamson/mammoth.js#33]
If you copy the following text into a .docx file and try to convert it, there is a "Maximum Call Stack Exceeded" error.

૱ Ă ă  0 1 2 3 4 5 6 7 8 9 Ǖ ǖ  ¤  Ð ¢ ℥ Ω ℧ K ℶ ℷ ℸ ⅇ ⅊  ᶀ ᶁ ᶂ ᶃ ᶄ ᶆ ᶇ ᶈ ᶉ ᶊ ᶋ ᶌ ᶍ ᶎ ᶏ ᶐ ᶑ ᶒ ᶓ ᶔ ᶕ ᶖ ᶗ ᶘ ᶙ ᶚ ᶸ ᵯ ᵰ ᵴ ᵶ ᵹ ᵼ ᵽ ᵾ ᵿ       ‎ ‏    ¥ £ ⅕ ⅙ ⅛ ⅔ ⅖ ⅗ ⅘ ⅜ ⅚  ⅝ ⅓  ⅞⍳ ⍴ ⍵ ⍶ ⍷ ⍸ ⍹ ⍺ » ‹ › ‘ ’ “ ” „ ‚ ❝ ❞ £ ¥ € $ ¢ ¬ ¶ @ § ® © ™ ° × π ± √ ‰ Ω ∞ ≈ ÷ ~ ≠ ¹ ² ³ ½ ¼ ¾ ‐ – — | ⁄ \ [ ] { } † ‡ … • • ● ⌃ ⇧↩ ¡ ¿ ‽α ß Á á À à Å å Ä ä Æ æ Ç ç É é È è Ê ê Í í Ì ì Î î Ñ ñ Ó ó Ò ò Ô ô Ö ö Ø ø Ú ú Ù ù Ü ü Ž ž ₳ ฿ ￠ € ₡ ¢ ₢ ₵ ₫ ￡ £ ₤ ₣ ƒ ₲ ₭ ₥ ₦ ₱ ＄ $ ₮ ₩ ￦ ¥ ￥ ₴ ₰ ¤ ៛ ₪ ₯ ₠ ₧ ₨ ௹ ﷼ ㍐ ৲ ৳ ~ ƻ Ƽ ƽ ¹ ¸ ¬ ¨ ɂ ǁ ¯ Ɂ ǂ ¡ ´ °  ¦ } { | . , • ] ) [ / _ \ ¿ º § " * - + ( ! & % $ ¼ ¾ ½ ¶ © ® @ `  ^ ꜠ ꜡  ' = : ; < ꞌ Ꞌ ꞊  ꞈ ꞉ > ? ÷ ℾ ℿ ℔ ℩ ℉ ⅀ ℈ þ ð Þ µ ª  ⱷ ⱶ Ⱶ ⱴ ⱱ  ⱦ ȶ ȴ ȣ Ȣ ȡ ȝ Ȝ ț ȋ Ȋ ȉ Ȉ ǯ Ǯ ǃ ǀ ƿ ƾ ƺ ƹ Ƹ Ʒ Ʋ ư ƪ ƣ Ƣ Ɵ ƛ Ɩ ƕ ƍ ſ  ₯ ἀ ἁ ἂ ἃ ἄ ἅ ἆ ἇ Ἀ Ἁ Ἂ Ἃ Ἄ Ἅ Ἆ Ἇ ἐ ἑ ἒ ἓ ἔ ἕ Ἐ Ἑ Ἒ Ἓ Ἔ Ἕ ἠ ἡ ἢ ἣ ἤ ἥ ἦ ἧ Ἠ Ἡ Ἢ Ἣ Ἤ Ἥ Ἦ Ἧ ἰ ἱ ἲ ἳ ἴ ἵ ἶ ἷ Ἰ Ἱ Ἲ Ἳ Ἴ Ἵ Ἶ Ἷ ὀ ὁ ὂ ὃ ὄ ὅ Ὀ Ὁ Ὂ Ὃ Ὄ Ὅ ὐ ὑ ὒ ὓ ὔ ὕ ὖ ὗ Ὑ Ὓ Ὕ Ὗ ὠ ὡ ὢ ὣ ὤ ὥ ὦ ὧ Ὠ Ὡ Ὢ Ὣ Ὤ Ὥ Ὦ Ὧ ὰ ά ὲ έ ὴ ή ὶ ί ὸ ό ὺ ύ ὼ ώ ᾀ ᾁ ᾂ ᾃ ᾄ ᾅ ᾆ ᾇ ᾈ ᾉ ᾊ ᾋ ᾌ ᾍ ᾎ ᾏ ᾐ ᾑ ᾒ ᾓ ᾔ ᾕ ᾖ ᾗ ᾘ ᾙ ᾚ ᾛ ᾜ ᾝ ᾞ ᾟ ᾠ ᾡ ᾢ ᾣ ᾤ ᾥ ᾦ ᾧ ᾨ ᾩ ᾪ ᾫ ᾬ ᾭ ᾮ ᾯ ᾰ ᾱ ᾲ ᾳ ᾴ ᾶ ᾷ Ᾰ Ᾱ Ὰ Ά ᾼ ᾽ ι ᾿ ῀ ῁ ῂ ῃ ῄ ῆ ῇ Ὲ Έ Ὴ Ή ῌ ῍ ῎ ῏ ῐ ῑ ῒ ΐ ῖ ῗ Ῐ Ῑ Ὶ Ί ῝ ῞ ῟ ῠ ῡ ῢ ΰ ῤ ῥ ῦ ῧ Ῠ Ῡ Ὺ Ύ Ῥ ῭ ΅ ` ῲ ῳ ῴ ῶ ῷ Ὸ Ό Ὼ Ώ ῼ ´ ῾  ʹ ͵  ͺ ͻ ͼ ͽ ; ΄ ΅ Ά · Έ Ή Ί Ό Ύ Ώ ΐ Α Β Γ Δ Ε Ζ Η Θ Ι Κ Λ Μ Ν Ξ Ο Π Ρ Σ Τ Υ Φ Χ Ψ Ω Ϊ Ϋ ά έ ή ί ΰ α β γ δ ε ζ η θ ι κ λ μ ν ξ ο π ρ ς σ τ υ φ χ ψ ω ϊ ϋ ό ύ ώ ϐ ϑ ϒ ϓ ϔ ϕ ϖ ϗ Ϙ ϙ Ϛ ϛ Ϝ ϝ Ϟ ϟ Ϡ ϡ Ϣ ϣ Ϥ ϥ Ϧ ϧ Ϩ ϩ Ϫ ϫ Ϭ ϭ Ϯ ϯ ϰ ϱ ϲ ϳ ϴ ϵ ϶ Ϸ ϸ Ϲ Ϻ ϻ ϼ Ͻ Ͼ Ͽ A a Ạ ạ Ả ả Ḁ ḁ Â Ã Ǎ ǎ Ấ ấ Ầ ầ Ẩ ẩ Ȃ ȃ Ẫ ẫ Ậ ậ À Á Ắ ắ Ằ ằ Ẳ ẳ Ẵ ẵ Ặ ặ Ā ā Ą ą Ǟ Ȁ ȁ Å Ǻ ǻ Ä ä ǟ Ǡ ǡ â á å ã à ẚ Ȧ ȧ Ⱥ Å ⱥ Æ æ Ǽ Ǣ ǣ  Ɑ  ª ℀ ⅍ ℁  B b Ḃ ḃ Ḅ ḅ Ḇ ḇ Ɓ Ƀ ƀ ƃ Ƃ Ƅ ƅ ℬ  C c Ḉ ḉ Ć ć Ĉ ĉ Ċ ċ Č č Ç ç Ƈ ƈ Ȼ ȼ ℂ ℃ ℭ Ɔ ℅ ℆ ℄  D d Ḋ ḋ Ḍ ḍ Ḏ ḏ Ḑ ḑ Ḓ ḓ Ď ď Ɗ Ƌ ƌ Ɖ Đ đ ȡ  Ǳ ǲ ǳ Ǆ ǅ ǆ ȸ E e Ḕ ḕ Ḗ ḗ Ḙ ḙ Ḛ ḛ Ḝ ḝ Ẹ ẹ Ẻ ẻ Ế ế Ẽ ẽ Ề ề Ể ể Ễ ễ Ệ ệ Ē ē Ĕ ĕ Ė ė Ę ę Ě ě È è É é Ê ê Ë ë Ȅ ȅ Ȩ ȩ Ȇ ȇ Ǝ  Ɇ ℇ ℯ ℮ Ɛ ℰ Ə ǝ  ɇ F f Ḟ ḟ Ƒ ƒ  Ⅎ ⅎ ℱ G g Ɠ Ḡ ḡ Ĝ ĝ Ğ ğ Ġ ġ Ģ ģ Ǥ ǥ Ǧ ǧ Ǵ ℊ ⅁ ǵ  H h Ḣ ḣ Ḥ ḥ Ḧ ḧ Ḩ ḩ Ḫ ḫ ẖ Ĥ ĥ Ȟ ȟ Ħ ħ Ⱨ ⱨ   Ƕ ℏ ℎ ℋ ℌ  I i Ḭ ḭ Ḯ ḯ Ĳ ĳ ì í î ï Ì Í Î Ï Ĩ ĩ Ī ī Ĭ ĭ Į į ı Ɨ ƚ  Ǐ ǐ  ℹ ℑ ℐ  J j Ĵ ĵ ȷ  Ɉ ɉ ǰ  K k Ḱ ḱ Ḳ ḳ Ḵ ḵ Ķ ķ Ƙ ƙ  Ǩ ǩ Ⱪ ⱪ ĸ  L l Ḷ ḷ Ḹ ḹ Ḻ ḻ Ḽ ḽ Ĺ ĺ Ļ ļ Ľ İ ľ Ŀ ŀ Ł ł Ỉ ỉ Ị ị Ƚ Ⱡ  ⱡ Ɫ  ℒ Ǉ ǈ ǉ ⅃ ⅂ ℓ ȉ Ȉ Ȋ ȋ  M m Ḿ ḿ Ṁ ṁ Ṃ ṃ  Ʃ Ɯ ℳ  N n Ṅ ṅ Ṇ ṇ Ṉ ṉ Ṋ ṋ Ń ń Ņ ņ Ň ň Ǹ ǹ Ŋ Ɲ ñ ŉ Ñ Ƞ ƞ ŋ Ǌ ǋ ǌ ȵ  №   O o Ṍ ṍ Ṏ ṏ Ṑ ṑ Ṓ ṓ Ȫ ȫ Ȭ ȭ Ȯ ȯ Ȱ ȱ Ǫ ǫ Ǭ ǭ Ọ ọ Ỏ ỏ Ố ố Ồ ồ Ổ ổ Ỗ ỗ Ộ ộ Ớ ớ Ờ ờ Ở ở Ỡ ỡ Ợ ợ Ơ ơ Ō ō Ŏ ŏ Ő ő Ò Ó Ô Õ Ö Ǒ Ȍ ȍ Ȏ ȏ Œ œ Ø Ǿ  ǽ ǿ ℴ ⍥ ⍤  ò ó ô õ ö ǒ ø  P p Ṕ ṕ Ṗ ṗ Ƥ ƥ Ᵽ ℙ Ƿ ℘ Q q Ɋ ɋ ℚ ℺ ȹ R r Ŕ ŕ Ŗ ŗ Ř ř Ṙ ṙ Ṛ ṛ Ṝ ṝ Ṟ ṟ Ȑ ȑ Ȓ ȓ ɍ Ɍ Ʀ Ɽ ℞  ℜ ℛ ℟ S s Ṡ ṡ Ṣ ṣ Ṥ ṥ Ṧ ṧ Ṩ ṩ Ś ś Ŝ ŝ Ş ş Š š Ș ș ȿ Ƨ ƨ ẞ ß ẛ  ℠ T t Ṫ ṫ Ṭ ṭ Ṯ ṯ Ṱ ṱ Ţ ţ Ť ť Ŧ ŧ Ƭ Ʈ ẗ Ț Ⱦ ƫ ƭ ț ⱦ ȶ ™ U u Ṳ ṳ Ṵ ṵ Ṷ ṷ Ṹ ṹ Ṻ ṻ Ụ Ủ ủ Ứ Ừ ụ ứ Ử ử ừ ữ Ữ Ự ự Ũ ũ Ū ū Ŭ ŭ Ů ů Ű ű Ǚ ǚ Ǘ ǘ Ǜ ǜ Ų ų Ǔ ǔ Ȕ ȕ Û û Ȗ ȗ Ù ù Ü ü Ư ú Ʉ ư Ʋ Ʊ  V v Ṽ ṽ Ṿ ṿ  Ʌ ℣ ⱱ ⱴ  W w Ẁ ẁ Ẃ ẃ Ẅ ẅ Ẇ ẇ Ẉ ẉ Ŵ ŵ ẘ Ⱳ ⱳ  X x Ẋ ẋ Ẍ ẍ ℵ ×  y Y Ẏẏ  ẙ Ỳ ỳ Ỵ ỵ Ỷ ỷ Ỹ ỹ Ŷ ŷ Ƴ ƴ Ÿ ÿ Ý ý Ɏ ɏ Ȳ Ɣ ⅄ ȳ  Z z Ẑ ẑ Ẓ ẓ Ẕ ẕ Ź ź Ż ż Ž ž Ȥ ȥ Ⱬ ⱬ Ƶ ƶ ɀ ℨ ℤ  ∀ ∁ ∂ ∃ ∄ ∅ ∆ ∇ ∈ ∉ ∊ ∋ ∌ ∍  ∏ ∐ ∑ ⋲ ⋳ ⋴ ⋵ ⋶ ⋷ ⋸ ⋹ ⋺ ⋻ ⋼ ⋽ ⋾ ⋿  € ₤ ＄ Ω β Φ Σ Ξᶛ ᶜ ᶝ ᶞ ᶟ ᶠ ᶡ ᶢ ᶣ ᶤ ᶥ ᶦ ᶧ ᶨ ᶩ ᶪ ᶫ ᶬ ᶭ ᶮ ᶯ ᶰ ᶱ ᶲ ᶳ ᶴ ᶵ ᶶ ᶷ ᶹ ᶺ ᶻ ᶼ ᶽ ᶾ ᶿ ᴀ ᴁ ᴂ ᴃ ᴄ ᴅ ᴆ ᴇ ᴈ ᴉ ᴊ ᴋ ᴌ ᴍ ᴎ ᴏ ᴐ ᴑ ᴒ ᴓ ᴔ ᴕ ᴖ ᴗ ᴘ ᴙ ᴚ ᴛ ᴜ ᴝ ᴞ ᴟ ᴠ ᴡ ᴢ ᴣ ᴤ ᴥ ᴦ ᴧ ᴨ ᴩ ᴪ ᴫ ᴬ ᴭ ᴮ ᴯ ᴰ ᴱ ᴲ ᴳ ᴴ ᴵ ᴶ ᴷ ᴸ ᴹ ᴺ ᴻ ᴼ ᴽ ᴾ ᴿ ᵀ ᵁ ᵂ ᵃ ᵄ ᵅ ᵆ ᵇ ᵈ ᵉ ᵊ ᵋ ᵌ ᵍ ᵎ ᵏ ᵐ ᵑ ᵒ ᵓ ᵔ ᵕ ᵖ ᵗ ᵘ ᵙ ᵚ ᵛ ᵜ ᵝ ᵞ ᵟ ᵠ ᵡ ᵢ ᵣ ᵤ ᵥ ᵦ ᵧ ᵨ ᵩ ᵪ ᵫ ᵬ ᵭ ᵮ ᵱ ᵲ ᵳ ᵵ ᵷ ᵸ ᵺ ᵻ  ‘ ’ ‛ ‚ “ ” „ ‟ « » ‹ › Ꞌ " ❛ ❜ ❝ ❞ < > @ ‧ ¨ ․ ꞉ :  ⁞ ‥ …  ; ― _  ¬ / \ ⁄ \ ⁄ | ⎜ ¦ ‖ ‗ † ‡ • •  ° ‣ % ‰ ‱ & ⅋ § ÷ + ± = ꞊ ′ ″ ‴ ⁗ ‵ ‶ ‷ ‸ * ⁑ ! ‼ ¡ ? ¿ ⁇ ⁉ ⁈ ‽ ¼ ½ ¾ ² ³ © ® ™ ℠  ℅ ℁ ⅍ ℄

**Comment by mwilliamson** [comment]
I couldn't reproduce when trying it myself on node-webkit, so closing. Feel free to reopen if this is still an issue.

**Comment by jagi** [comment]
[mammoth_error.docx](https://github.com/mwilliamson/mammoth.js/files/289634/mammoth_error.docx)
[mammoth_ok.docx](https://github.com/mwilliamson/mammoth.js/files/289635/mammoth_ok.docx)

I think that I found source of the problem. It's not some character that is causing a problem but rather too many words in one paragraphs. I haven't been investigating mammoth's code but I assume that it may be calling recursively parsing function for each word or sth like that. However, let's take a look at the example documents I've uploaded. The only difference between them is that the working one has extra RETURN character in the middle of the paragraph. Without that it's not working. I've been receiving two different errors with different sample files.

```sh
W20160530-14:56:10.654(2)? (STDERR) /Users/jagi/workspace/usefulio/parley/node_modules/mammoth/lib/html/ast.js:0
W20160530-14:56:10.654(2)? (STDERR) (function (exports, require, module, __filename, __dirname) { var _ = require(
W20160530-14:56:10.654(2)? (STDERR) ^
W20160530-14:56:10.656(2)? (STDERR) RangeError: Maximum call stack size exceeded
```

```sh
W20160530-15:11:23.623(2)? (STDERR) /Users/jagi/workspace/usefulio/parley/node_modules/mammoth/lib/document-to-html.js:105
W20160530-15:11:23.624(2)? (STDERR)         convertElements(run.children, messages, function(error, children) {
W20160530-15:11:23.624(2)? (STDERR)                                                         ^
W20160530-15:11:23.626(2)? (STDERR) RangeError: Maximum call stack size exceeded
```

@mwilliamson what do you think about it?

**Comment by mwilliamson** [comment]
Thanks for the document, I can reproduce the error with that. It looks the problem is that a single paragraph has many children, which (due to the way that `async.mapSeries` works) causes the stack overflow. Specifically, it's probably [this call](https://github.com/mwilliamson/mammoth.js/blob/c6bb2dfe79ed6fb1efe5e9a385f61e78d1224765/lib/document-to-html.js#L58).

There's already some mitigation in there with the call to `setImmediate`, but that only happens between paragraphs -- it doesn't help once you're in a paragraph. The simplest solution is probably to add something similar more consistently, but taking care not to significantly degrade performance. I'll take a look when I get a chance, but feel free to take a look.

**Comment by jagi** [comment]
After reading documentation of the `async` package I think that solution for this problem would be using the `async.ensureAsync` util function.
```js
async.mapSeries(elements, async.ensureAsync(function(element, callback) {
  return elementToHtml(element, messages, callback);
}), callback);
```
However, `mammoth.js` is using very old version (0.2.6) of the `async` package and there is no such function. And the current stable version is 1.5.2, and 2.0 is in the RC stage. I think you should consider upgrading it.

**Comment by mwilliamson** [comment]
I agree, but I'd be hesitant to use it until the effect on performance is known. (Performance is the reason why this was originally written using callbacks rather than promises, as the rest of Mammoth is. Having said that, it might well be the case that the performance of promises has improved in both node.js and the browser sufficiently that promises can be used instead.)

**Comment by jagi** [comment]
@mwilliamson I'm not an expert in that field but from what I've read bluebird is faster than native promises and async.

**Comment by jagi** [comment]
@mwilliamson as we talk about speed of conversion. I've just measured conversion of the same document using Mammoth.js and CKEditor just by pasting content from the docx file. It took 2 seconds CKEditor to convert document to HTML, and Mammoth.js finished task in 13 seconds. The only disadvantage of CKEditor is that it produces a little bit worse formatted HTML than Mammoth.js. I'm just curious whys is Mammoth.js around 6 time slower? Can some optimisation be made to make it faster?

**Comment by mwilliamson** [comment]
Knowing nothing about how CKEditor works, I have no idea why it's faster. Pull requests with performance improvements are welcome though (so long as it doesn't adversely affect code clarity)!

**Comment by mwilliamson** [comment]
> @mwilliamson I'm not an expert in that field but from what I've read bluebird is faster than native promises and async.

I think actually trying it is the only way to be sure, both in node.js and the browser.

**Comment by jagi** [comment]
Apparently it's not a bug in `mammoth` but a bug in `async` library. I've just updated it to the newest stable version (1.5.2) and it started working. All my test documents now are properly converted. Can you confirm that and merge my PR?

**Comment by mwilliamson** [comment]
I believe async 2 is getting rid of the `setImmediate` calls, so I'd rather have a solution that works with it. It's probably also worth checking that the stack size isn't being reduced by a constant factor i.e. that more children won't cause a stack overflow.

**Comment by jagi** [comment]
I can make some tests with number of children but if you want to switch to async 2, then when can we expect that to happen?

**Comment by mwilliamson** [comment]
That depends on when the async project is planning to release version 2 (which I don't know). I don't think we need to wait to upgrade to async 2, but we would preferable end up with something that works with it.

A viable alternative might be to make the conversion to HTML synchronous, and insert promises into the output where we currently would otherwise perform a synchronous operation. We can then evaluate all of the promises and produce a new HTML AST asynchonously -- in other words, batching up all of the asynchronous operations, which should be overwhelming smaller than the total of elements.

**Comment by mwilliamson** [comment]
The branch `conversion-sync-phase` handles asynchrony in the conversion by batching up the asynchronous actions until the end. Would you mind giving it a go and see if it works for you?

**Comment by mwilliamson** [comment]
This should be fixed by the latest release (1.0.4), let me know if not.

**Comment by jagi** [comment]
Ok thanks, I will check it out and let you know
